src/common/latest/prepare-task.ts

```diff
@@ -41,33 +46,72 @@ export const prepareTask: TaskJob = async (endpointType: EndpointType) => {
     ...parseScannerExtraProperties(),
   };
 
-  tl.setVariable(TaskVariables.SonarQubeServerVersion, serverVersion.format());
-  tl.setVariable(TaskVariables.SonarQubeScannerMode, scannerMode);
-  tl.setVariable(
-    TaskVariables.SonarQubeScannerReportTaskFile,
-    props["sonar.scanner.metadataFilePath"],
-  );
-  tl.setVariable(TaskVariables.SonarQubeEndpoint, endpoint.toJson(), true);
+  tl.setVariable(TaskVariables.SonarServerVersion, serverVersion.format());
+  tl.setVariable(TaskVariables.SonarScannerMode, scannerMode);
+  tl.setVariable(TaskVariables.SonarScannerReportTaskFile, props["sonar.scanner.metadataFilePath"]);
+  tl.setVariable(TaskVariables.SonarEndpoint, endpoint.toJson(), true);
+  tl.setVariable(TaskVariables.SonarMsBuildVersion, msBuildVersion);
+  tl.setVariable(TaskVariables.SonarCliVersion, cliVersion);
 
   tl.getDelimitedInput("extraProperties", "\n")
     .filter((keyValue) => !keyValue.startsWith("#"))
     .map((keyValue) => keyValue.split(/=(.+)/))
     .forEach(([k, v]) => (props[k] = v));
 
-  tl.setVariable(TaskVariables.SonarQubeScannerMode, scannerMode);
-  tl.setVariable(TaskVariables.SonarQubeEndpoint, endpoint.toJson(), true);
+  tl.setVariable(TaskVariables.SonarScannerMode, scannerMode);
+  tl.setVariable(TaskVariables.SonarEndpoint, endpoint.toJson(), true);
 
   const jsonParams = stringifyScannerParams({
     ...endpoint.toSonarProps(serverVersion),
     ...scanner.toSonarProps(),
     ...props,
   });
 
-  tl.setVariable(TaskVariables.SonarQubeScannerParams, jsonParams);
+  tl.setVariable(TaskVariables.SonarScannerParams, jsonParams);
+
+  // always download the scanner, uses default versions if not explicitly overriden
+  const scannerPath = await downloadScanner(cliVersion, msBuildVersion);
 
+  tl.setVariable(TaskVariables.SonarScannerLocation, scannerPath);
+
+  // we need to pass in the downloaded paths here
   await scanner.runPrepare();
 };
 
+async function downloadScanner(cliVersion: string, msBuildVersion: string) {
+  const scannerMode = tl.getVariable(TaskVariables.SonarScannerMode);
+
+  let fileUrl;
+  // Override the default URL if a version differs from the default
+  if (scannerMode === ScannerMode.CLI) {
+    fileUrl = scannerConfig.cliUrlTemplate(cliVersion);
+  } else if (scannerMode === ScannerMode.MSBuild) {
+    fileUrl = scannerConfig.msBuildUrlTemplate(msBuildVersion, isWindows());
+  }
```

Comment:
In case of gradle/maven scanner, the task would fail right? Because then `toolLib.downloadTool()` is called with undefined?

Reply:
thats true, probably should have a test for this 😅